source/opt/dead_branch_elim_pass.cpp

```diff
@@ -0,0 +1,271 @@
+      *condVal = false;
+      *condIsConst = true;
+    } break;
+    case SpvOpConstantTrue: {
+      *condVal = true;
+      *condIsConst = true;
+    } break;
+    case SpvOpLogicalNot: {
+      bool negVal;
+      (void)GetConstCondition(cInst->GetSingleWordInOperand(0),
+          &negVal, condIsConst);
+      if (*condIsConst)
+        *condVal = !negVal;
+    } break;
+    default: {
+      *condIsConst = false;
+    } break;
+  }
+}
+
+void DeadBranchElimPass::AddBranch(uint32_t labelId, ir::BasicBlock* bp) {
+  std::unique_ptr<ir::Instruction> newBranch(
+    new ir::Instruction(SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {labelId}}}));
+  def_use_mgr_->AnalyzeInstDefUse(&*newBranch);
+  bp->AddInstruction(std::move(newBranch));
+}
+
+void DeadBranchElimPass::KillAllInsts(ir::BasicBlock* bp) {
+  bp->ForEachInst([this](ir::Instruction* ip) {
+    def_use_mgr_->KillInst(ip);
+  });
+}
+
+bool DeadBranchElimPass::GetConstConditionalBranch(ir::BasicBlock* bp,
+    ir::Instruction** branchInst, ir::Instruction** mergeInst,
+    bool *condVal) {
+  auto ii = bp->end();
+  --ii;
+  *branchInst = &*ii;
+  if ((*branchInst)->opcode() != SpvOpBranchConditional)
+    return false;
+  --ii;
+  *mergeInst = &*ii;
+  if ((*mergeInst)->opcode() != SpvOpSelectionMerge)
+    return false;
+  bool condIsConst;
+  (void) GetConstCondition(
+      (*branchInst)->GetSingleWordInOperand(kBranchCondConditionalIdInIdx),
+      condVal, &condIsConst);
+  return condIsConst;
+}
+
+bool DeadBranchElimPass::EliminateDeadBranches(ir::Function* func) {
+  // Traverse blocks in structured order
+  std::list<ir::BasicBlock*> structuredOrder;
+  ComputeStructuredOrder(func, &structuredOrder);
+  std::unordered_set<ir::BasicBlock*> elimBlocks;
+  bool modified = false;
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    // Skip blocks that are already in the elimination set
+    if (elimBlocks.find(*bi) != elimBlocks.end())
+      continue;
+    // Skip blocks that don't have constant conditional branch
+    ir::Instruction* br;
+    ir::Instruction* mergeInst;
+    bool condVal;
+    if (!GetConstConditionalBranch(*bi, &br, &mergeInst, &condVal))
+      continue;
+
+    // Replace conditional branch with unconditional branch
+    const uint32_t trueLabId =
+        br->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx);
+    const uint32_t falseLabId =
+        br->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
+    const uint32_t mergeLabId =
+        mergeInst->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+    const uint32_t liveLabId = condVal == true ? trueLabId : falseLabId;
+    const uint32_t deadLabId = condVal == true ? falseLabId : trueLabId;
+    AddBranch(liveLabId, *bi);
+    def_use_mgr_->KillInst(br);
+    def_use_mgr_->KillInst(mergeInst);
+
+    // Iterate to merge block deleting dead blocks
+    std::unordered_set<uint32_t> deadLabIds;
+    deadLabIds.insert(deadLabId);
```

Comment:
There's a difficulty here because, I believe, the structured order of blocks could put a continue target we break out to  before an inner merge block.  (Example below)

So... When one of the candidates to kill is a "continue block", then it this could try to kill the corresponding continue target, and that would also orphan the reference in the corresponding OpLoopMerge.

Example is schematically similar to:
```
  for (...) {  // names a continue target
      if ( ... ) {
        break; // a branch to the continue target.
         // but the merge block for this structured selection will appear *after* the continue target in the structured order.
         // So we'll want to sweep away the continue target block when we shouldn't.
      }
     ; continue target is conceptually here.
  }
```

More explicitly here's a full example:

```
        OpCapability Shader
        OpMemoryModel Logical GLSL450
        OpEntryPoint Fragment %foo "foo"
%bool = OpTypeBool
%void = OpTypeVoid
%void_fn = OpTypeFunction %void
%boolptr = OpTypePointer Function %bool
%false = OpConstantFalse %bool

%foo =  OpFunction %void None %void_fn

%entry = OpLabel
%loopcondition_var = OpVariable %boolptr Function
%loopcondition = OpLoad %bool %loopcondition_var
OpBranch %loop

%loop = OpLabel
OpLoopMerge %mergeloop %continuetarget None
OpBranchConditional %loopcondition %mergeloop %body

%body = OpLabel
OpSelectionMerge %mergeselection None
OpBranchConditional %false %continuetarget %mergeselection

%mergeselection = OpLabel
OpKill

%continuetarget = OpLabel
OpBranchConditional %loopcondition %loop %mergeloop

%mergeloop = OpLabel
OpReturn
OpFunctionEnd
     
```
The above example passes validation.  Running it through your algorithm, I get this result:
```
               OpCapability Shader
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %1 "foo"
       %bool = OpTypeBool
       %void = OpTypeVoid
          %4 = OpTypeFunction %void
%_ptr_Function_bool = OpTypePointer Function %bool
      %false = OpConstantFalse %bool
          %1 = OpFunction %void None %4
          %7 = OpLabel
          %8 = OpVariable %_ptr_Function_bool Function
          %9 = OpLoad %bool %8
               OpBranch %10
         %10 = OpLabel
               OpLoopMerge %11 %12 None ;;;;;;;;;;;;;;;; Whoops this is a dead reference to %12, and is invalid.
               OpBranchConditional %9 %11 %13
         %13 = OpLabel
               OpBranch %14
         %14 = OpLabel
               OpKill
         %11 = OpLabel
               OpReturn
               OpFunctionEnd
```